### DISCUSSION TITLE
Connection Lora 1278 to Arduino

### DISCUSSION THREAD

**Discussion by derungebundener** [jgromes/RadioLib#222] [Q&A]
Hello, i'v got a little question:

I want to connect my Lora 1278 to Arduino, code in exmaples says that we have to connect 4 wires:

**// NSS pin:   10
// DIO0 pin:  2
// RESET pin: 9
// DIO1 pin:  3**
And also +5V and GND

Before this library i was trying another:
https://github.com/sandeepmistry/arduino-LoRa

And there i found that we have to connect almost all wires:

**SCK toSCK
MISO to MISO
MOSI  to  MOSI
NSS to 10
NRESET to 9
DIO0 to 2**
And also +5V / GND

Can you tell me please about correct wiring? Thanks :)

P.S. English is'nt my first, sorry about mistakes.

**Comment by jgromes** [comment]
Since the LoRa module uses SPI to communicate, you have to connect all the SPI wires - the examples only show the pins that can be connected to any digital pin. On most Arduino boards (like Uno), SPI pins MISO/MOSI/SCK are fixed.

**Reply by derungebundener** [reply]
Thanks for answer!
If i understood correct - SPI are always fixed and in this exmaple we can free change those 4 pins, yes?

And another one question:
I have a little bit "special" LoRa module. It comes with antenna and amplifier 1W power that equal to 30dbm.

I was asking manufacturer about power chaning, he answered me with this picture:
https://i.imgur.com/j8Wir3e.jpg
If i'v understood right - we have SPI registers and must write them to their adress.

In your library - where can i find this piece of code than i should to rewrite with this "power registers"?

Thanks :)

**Reply by jgromes** [reply]
> If i understood correct - SPI are always fixed and in this exmaple we can free change those 4 pins, yes?

In the most common cases (i.e. with Arduino Uno), yes. On some other platforms like STM32 or ESP32 you can change the SPI pins as well.

> I have a little bit "special" LoRa module

Which one? Does the SX1278 really have an option to control the *external* PA from its *internal* SPI registers? That seems unlikely. In any case I can't recommend to start poking around in the PA registers, that's definitely something you can't just write a random number to "because the seller says so". The library also avoids these sorts of undocumented changes because they can have quite profound consequences. Long story short, unless it comes from Semtech (the SX1278 chip manufacturer), I will be very hesitant to put it into the library.

**Reply by derungebundener** [reply]
I'v bought this clone from "nicerf"
https://www.nicerf.com/product_193_269.html
And asked them about power.

**Reply by derungebundener** [reply]
P.S. On their page in downloadable PDF-manual i'v found this:
https://i.imgur.com/UpYlKx3.png
I think seller "a little bit" right - we can change power with those registers...

**Reply by jgromes** [reply]
Looking over their docs it looks like it's just a simple external PA, so it probably just amplifies whatever is at the input - same as other 30 dBm modules, like for example Ebyte's E22400M30S, it's based on SX1268 LoRa chip with YP2233W external PA with 26 dB gain.

The library already takes care of this sort of configuration - see the `setOutputPower()` function. The screenshot you sent before also sets some other bits in other registers, with no explanation why, that's what I would be cautious about.

BTW do you have a license? 1W transmitted power is most likely well above license-free operation in your country.

**Reply by derungebundener** [reply]
> BTW do you have a license?

Im from Russia - 433mhz if free here :)

**Comment by N6RFM** [comment]
I have been experimenting with the GNICE sx1268F30.  P0 may be modified by varying the drive level. 

Users may also find this breakout of interest, since it appears that the chips share the same spacing.  Interestingly, the sx1278F30 does not explicitly label BUSY, as compared to the sx1268F30, but denote DIO0 that physical location.

And yes, in the USA (and many other countries) transmitting in the 70cm amateur radio band requires a license, at any power out level!  

https://github.com/N6RFM/LoRA-PCB

Thanks Jan for great support!  
73,
Bob
N6RFM

**Reply by jgromes** [reply]
> I have been experimenting with the GNICE sx1268F30. P0 may be modified by varying the drive level.

Do you have some information on the PA it's using? E22400M30S has 26 dB PA, so to get the 30 dBm output I'd feed it 4 dBm input if I'm not mistaken. Not sure what's the situation with GNice modules.

>Interestingly, the sx1278F30 does not explicitly label BUSY, as compared to the sx1268F30

I would expect so, since SX1278 does not have a BUSY pin ;)

**Reply by jgromes** [reply]
@N6RFM great minds think alike I see :) https://github.com/jgromes/ESP32-1W-GroundStation

**Reply by derungebundener** [reply]
> P0 may be modified

What it means? Some kind of "pin 0" or word "power"?

**Reply by N6RFM** [reply]
Sorry Jan no information on PA inside the GNICE chips.  

Fantastic job on your eByte board.  WOW.  Really like the design!  :-)

**Reply by jgromes** [reply]
@derungebundener I believe @N6RFM meant the output power there. In short you'll need to figure out what is the PA inside the GNice module, find its datasheet and from that find out its gain. Then you will know the relation between SX1278 output power and the PA output power. For example, if the PA has 26 dB gain, you will have to call `setOutputPower(4)` to get 30 dBm of actual output.

**Reply by derungebundener** [reply]
Thanks for helping guys!
I'll try it tomorrow and answer you about results - its already 2:40 hours in Russia.

Btw, its it possible with this library to read incoming power on reciever? I was thinking - i can try put sender's and reciever's antenas close to each other and look at inkoming power. In another library it calls like "RSSI" or something near - not sure...

Thanks one more time! :)

**Reply by jgromes** [reply]
You can read RSSI (Recorded Signal Strength Indicator) with `getRSSI()`, SNR (Signal-to-Noise Ratio) with `getSNR()` and the received signal frequency error with `getFrequencyError()`. I suggest checking out the SX127x examples (https://github.com/jgromes/RadioLib/blob/master/examples/SX127x).

**Reply by derungebundener** [reply]
Sorry for delay - i'll answer about result later, already 3 hours f***ing with logick-level-converter between 5 and 3V. So many wires to solder :(

**Comment by derungebundener** [comment]
Finaly i'v soldered first LoRa and tried "chanel activity detection" and "recive" - this library works good, thanks :)

Can you help please with another question?
How can i change default settings for LoRa 1278 .begin function?
Yes, i know -  can change it via **.begin (bla-bla-bla)** but in my project i would use usual **.begin** for all situations.

Thanks again ans sorry for disturbing :)

**Reply by jgromes** [reply]
I suppose if you really wanted to change the default settings, you'd go to src/modules/SX127x/SX127x.h and change it there, but why would you?

If you want to change just one of the settings and keep the rest as-is, you can use one of the configuration methods (see the example SX127x_Settings and the API documentation). If you want to change a lot of them, it's easier to specify it all in the begin method, like you said.

**Reply by derungebundener** [reply]
Thanks!

Yes, it looks a little bit strange and un-normal, but i would use my setting as default - so i will be sure that i didnt any mistake in settings :)

**Reply by jgromes** [reply]
Surely that will be more prone to an error - you will have to keep using that modified version of the library, leaving you stuck with the current release, or always remember to redo the change after updating the library.

**Reply by derungebundener** [reply]
Yes, i know about it - its not a big problem.

**Comment by derungebundener** [comment]
Hello again! I dont want to start new theme, so i will ask in this...

I want try t use NRFl2401 module. Im beginner in arduino-theme and heared thet we can you "acknoledge" (or something like that) with usefull payload. I think it's some kind of confirmation, yes? So, can we with this liblary to pack usefull data and send them in this confirmation message? Thanks )

P.S. As i said before - english isnt my first, so ims sorry about mistakes ;-)
